pkg/controller/petset/iterator.go

```diff
@@ -0,0 +1,168 @@
+package petset
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+// petGenerator generates pets.
+type petGenerator struct {
+	// ps is the petset for which this generator generates pets
+	ps *apps.PetSet
+	// idMappers are a list of identityMapper interfaces capable of giving
+	// the generates pets an identity
+	idMappers []identityMapper
+}
+
+// newPet generates a new pet using the id string as a unique qualifier
+func (pg *petGenerator) newPet(id string) (*pcb, error) {
+	petPod, err := controller.GetPodFromTemplate(&pg.ps.Spec.Template, pg.ps)
+	if err != nil {
+		return nil, err
+	}
+	for _, im := range pg.idMappers {
+		im.AddIdentity(id, petPod)
+	}
+	petPVCs := []api.PersistentVolumeClaim{}
+	for _, c := range GetClaims(id, pg.ps) {
+		petPVCs = append(petPVCs, c)
+	}
+	// TODO: Replace id field with IdentityHash, since id is more than just an index.
+	return &pcb{pod: petPod, pvcs: petPVCs, id: id, parent: pg.ps}, nil
+}
+
+// NewPetGenerator returns a pet generator for the given petset
+func NewPetGenerator(ps *apps.PetSet) *petGenerator {
+	return &petGenerator{
+		ps: ps,
+		idMappers: []identityMapper{
+			&NameIdentityMapper{ps},
+			&VolumeIdentityMapper{ps},
+			&NetworkIdentityMapper{ps},
+		},
+	}
+}
+
+// petQueue maintains a queue of pets
+type petQueue struct {
+	pets     []*pcb
+	idMapper identityMapper
+}
+
+// enqueue enqueues the given pet, evicting any pets with the same id
+func (pt *petQueue) enqueue(p *pcb) {
+	if p == nil {
+		pt.pets = append(pt.pets, nil)
+		return
+	}
+	// Pop an existing pet from the know list, append the new pet to the end.
+	petList := []*pcb{}
+	petID := pt.idMapper.GetIdentity(p.pod)
+	for i := range pt.pets {
+		if petID != pt.idMapper.GetIdentity(pt.pets[i].pod) {
+			petList = append(petList, pt.pets[i])
+		}
+	}
+	pt.pets = petList
+	p.event = syncPet
+	pt.pets = append(pt.pets, p)
+}
+
+// dequeue returns the last element of the queue
+func (pt *petQueue) dequeue() *pcb {
+	if pt.empty() {
+		glog.Warningf("Dequeue invoked on an empty queue")
+		return nil
+	}
+	l := len(pt.pets) - 1
+	pet := pt.pets[l]
+	pt.pets = pt.pets[:l]
+	return pet
+}
+
+func (pt *petQueue) empty() bool {
+	return len(pt.pets) == 0
+}
+
+// NewPetQueue returns a queue for tracking pets
+func NewPetQueue(ps *apps.PetSet, podList []*api.Pod) *petQueue {
+	pt := petQueue{pets: []*pcb{}, idMapper: &NameIdentityMapper{ps}}
+	// Seed the queue with existing pets. Assume all pets are scheduled for
```

Comment:
it's more a custom data structure that's functionally closes to a queue in behavior (we already have a queue and I would've just used that otherwise).